src/Rule/Ip.php

```diff
@@ -71,135 +46,181 @@ final class Ip extends Rule
+         * used, but the CIDR prefix is not set.
+         *
+         * You may use the following placeholders in the message:
+         *
+         * - `{attribute}`: the label of the attribute being validated
+         * - `{value}`: the value of the attribute being validated
+         *
+         * @see $allowSubnet
+         */
+        private string $noSubnetMessage = 'Must be an IP address with specified subnet.',
+        /**
+         * @var string user-defined error message is used when validation fails
+         * due to {@see $allowSubnet} is false, but CIDR prefix is present.
+         *
+         * You may use the following placeholders in the message:
+         *
+         * - `{attribute}`: the label of the attribute being validated
+         * - `{value}`: the value of the attribute being validated
+         *
+         * @see $allowSubnet
+         */
+        private string $hasSubnetMessage = 'Must not be a subnet.',
+        /**
+         * @var string user-defined error message is used when validation fails due to IP address
+         * is not allowed by {@see $ranges} check.
+         *
+         * You may use the following placeholders in the message:
+         *
+         * - `{attribute}`: the label of the attribute being validated
+         * - `{value}`: the value of the attribute being validated
+         *
+         * @see $ranges
+         */
+        private string $notInRangeMessage = 'Is not in the allowed range.',
+        /**
+         * @var array The IPv4 or IPv6 ranges that are allowed or forbidden.
+         *
+         * The following preparation tasks are performed:
+         *
+         * - Recursively substitutes aliases (described in {@see $networks}) with their values.
+         * - Removes duplicates.
+         *
+         * When the array is empty, or the option not set, all IP addresses are allowed.
+         *
+         * Otherwise, the rules are checked sequentially until the first match is found.
+         * An IP address is forbidden, when it has not matched any of the rules.
+         *
+         * Example:
+         *
+         * ```php
+         * (new Ip(ranges: [
+         *     '192.168.10.128'
+         *     '!192.168.10.0/24',
+         *     'any' // allows any other IP addresses
+         * ]);
+         * ```
+         *
+         * In this example, access is allowed for all the IPv4 and IPv6 addresses excluding the `192.168.10.0/24`
+         * subnet. IPv4 address `192.168.10.128` is also allowed, because it is listed before the restriction.
+         */
+        private array $ranges = [],
+        ?FormatterInterface $formatter = null,
+        bool $skipOnEmpty = false,
+        bool $skipOnError = false,
+        $when = null
+    ) {
+        parent::__construct(formatter: $formatter, skipOnEmpty: $skipOnEmpty, skipOnError: $skipOnError, when: $when);
+
+        foreach ($networks as $key => $_values) {
```

Comment:
It is differ from previous logic when it compared with both default and "new" networks. 
And about the test cases where I commented. Looks they will not pass now. Isn't it?

Reply:
With `network()` chained method calls it made sense. But with one array it will lead to duplicated array keys (PHP behavior in this case is the latter value will override the former without error), which I think is up to developer / IDE to check. Collisions with defaults are still handled though.

Reply:
:white_check_mark: